### ISSUE TITLE
Installation Problems

### ISSUE THREAD

**Issue by GTMux** [ammolytics/projects#28]
This project is great! And I would love to bring it in our shooting club for these people who like accurate reloads. Unfortunately I'm not able to get the RPi Zero running due to problems during the installation process:

.
.
.
.

npm install

> bluetooth-hci-socket@0.5.1 install /home/pi/node_modules/bluetooth-hci-socket
> node-gyp rebuild

make: Entering directory '/home/pi/node_modules/bluetooth-hci-socket/build'
  CXX(target) Release/obj.target/binding/src/BluetoothHciSocket.o
In file included from ../src/BluetoothHciSocket.cpp:8:
../../nan/nan.h: In function ‘void Nan::AsyncQueueWorker(Nan::AsyncWorker*)’:
../../nan/nan.h:2298:62: warning: cast between incompatible function types from ‘void (*)(uv_work_t*)’ {aka ‘void (*)(uv_work_s*)’} to ‘uv_after_work_cb’ {aka ‘void (*)(uv_work_s*, int)’} [-Wcast-function-type]
     , reinterpret_cast<uv_after_work_cb>(AsyncExecuteComplete)
                                                 
...
...
...
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:194:23)
gyp ERR! stack     at ChildProcess.emit (events.js:198:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:248:12)
gyp ERR! System Linux 4.19.75+
gyp ERR! command "/usr/bin/node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/pi/node_modules/bluetooth-hci-socket
gyp ERR! node -v v10.17.0
gyp ERR! node-gyp -v v5.0.5
gyp ERR! not ok

**Comment by erichiggins** [comment]
@GTMux Thanks for reporting this!  If you have the complete installation log available, would you mind including it here in the ticket, or as a separate Gist? I think some critical information from the trace may be missing.

Lastly, can you provide a little bit of detail about your setup?

- Host machine (Mac, Windows, Linux):
- Raspbian version:

Thanks again!

**Comment by GTMux** [comment]
Hi @erichiggins, 
thanks for the fast response. Setup is: 

- a Raspberry Zero W (armv6l)
- installed from an iMac (High Sierra / 10.13.6) using balenaEtcher to transfer the Raspbian image  (Raspbian Buster Lite, Version:September 2019, Release date:2019-09-26, Kernel version:4.19)

System setup started based on your description (rename, expand etc., update, install bluetooth bluez libbluetooth-dev libudev-dev, cloning project)

And than this time it seems to work fine except some warnings concerning missing pre-build resources (please see InstallLog.txt attached). Only the service installation ended with a failure (Log attached). And unfortunately I got an error message in the app on my android tab (SM-T210 Android Version 4.4.2) that no device is connected.

The difference between today and the last 5 days of installation trials: I only did:
sudo apt-get update 
and NOT sudo apt-get upgrade 
and NOT sudo apt-get distr-upgrade

Maybe you have some hints for me to get this running?

Thanks for your time!

Ralf

[2019-10-26_InstallLog.txt](https://github.com/ammolytics/projects/files/3774511/2019-10-26_InstallLog.txt)
[2019-10-26_ServiceInstfail.txt](https://github.com/ammolytics/projects/files/3774520/2019-10-26_ServiceInstfail.txt)

**Comment by GTMux** [comment]
Anyone have any ideas?
I would be very happy about advice. :) 

Thanks and best regards,
Ralf

**Comment by erichiggins** [comment]
Sorry for the delay (I was traveling for work)!

I have some time this weekend to dig into this issue -- I'd first like to see if I can reproduce it.

The real solution, I believe, would be for me to provide a ready-to-flash image that has everything installed. That avoids this type of complication.

**Comment by GTMux** [comment]
Thanks a lot - this would be great. I was close to ask you for an image. :)
Sorry to bother you with this problem...

**Comment by erichiggins** [comment]
Just a quick update on this: I have made some significant progress on a pre-configured image. It's nearly there, just requires a few more tweaks. Once it's ready, I'll share a link for you to give it a try. If it works for you, I'll update this repo and docs w/ an official link to it.

**Comment by erichiggins** [comment]
Another update: I know it's taking a while -- sorry about that! It's **very** close now.

More context: I'm using a system called [buildroot](https://buildroot.org/) to build a custom SD card image. [The manual is long and complex](https://buildroot.org/downloads/manual/manual.html), and it takes a few hours to run a complete build from scratch. So, it's not quite as easy as making a small change then being able to test it quickly. I still have to write the image to an SD card, boot a system, and see if everything is working.

It's coming along, but is very, very time-consuming. The two pieces I'm blocked on are:

1. Getting bluetooth working correctly
2. Getting the Open Trickler software to run correctly on boot

Thanks for your patience on this. If you happen to have any experience with buildroot, help is always appreciated!

**Comment by GTMux** [comment]
Hi Eric,

thanks for the update. I keep my fingers crossed that you succeed. 
Unfortunately, I have no idea how I could help with software related stuff...

**Comment by erichiggins** [comment]
Great news! I just got bluetooth working!
It's late now, so I haven't tried loading the OpenTrickler software yet. I'll give that a shot tomorrow -- it's looking very promising so far!

**Comment by erichiggins** [comment]
SUCCESS! I'll share an image file today along with some instructions. It'll help if someone can try it out and make sure it works on their end before I publish it to the world.

**Comment by GTMux** [comment]
great - would be a pleasure for me to test it! :)

**Comment by erichiggins** [comment]
[Test image file can be downloaded here](https://drive.google.com/open?id=1uTVCknwQI6sZAX_cFkh84gF3V91xCv71). Please be sure to read all of the following instructions and notes.

Please let me know if you have any questions, have any issues, or (hopefully) are able to get it working!

## Instructions

- Download the attached SD card image file  
  shasum: `47910ee0389418c11f7afd112ec4d9cad7d5ce06`
- Flash the image to your microSD card (I highly recommend [balena etcher](https://www.balena.io/etcher/))
- If you want it to be able to login via SSH over WiFi (optional):  
  Edit `boot/wpa_supplicant.conf` and add the SSID and password for your wireless network.
- Insert the microSD card into your Pi, connect it to your scale, and boot it up
- After about 60 seconds, try to connect to it with the [OpenTrickler Android App](https://play.google.com/store/apps/details?id=com.ammolytics.opentrickler&hl=en_US).

## Notes

- It's designed to use the PWM pin **`12`** on the Pi, rather than pin `15` which I used originally. You can change this in the `ecosystem.config.js` file on the `code` partition if needed (more below).
- It takes about 60-90 seconds for the system to fully boot and the OpenTrickler to be visible to the Android App.
- The microSD card will have three partitions: `boot`, `code`, and `opentrickler`. The first two are formatted with FAT32 and can be viewed/modified from Windows and Mac computers.
- The source code from this repo and all dependencies are contained on the `code` partition, so you can modify easily and try your own version.
- There are no logs available in this test image. I'm working on that.
- The attached file is compressed with [`xz`](https://en.wikipedia.org/wiki/XZ_Utils) to reduce the `.img` file from 320MB to 52MB. You *might* need to decompress it before flashing.
- It will not show up as `opentrickler.local` on your network. You'll need to check your WiFi router to figure out it's IP address in order to SSH into it.
- You can SSH into it using `root` as the login. The password is `ammolytics`
- There are some security risks associated with having a device connected to your network that has it's root password published online. I don't recommend using this image long-term.

**Comment by erichiggins** [comment]
Quick update: I'm working on another image that addresses some of the missing features that I listed above. Namely:

- Addressable on the network at `opentrickler.local`
- Improved startup performance
- Logging (if connected to the device via SSH)

**Comment by GTMux** [comment]
Hi Eric,

Thanks for your updates. I downloaded the image, decompressed it, brought it on the SD card using balena etcher with success. Due to the fact that my scale is currently borrowed by a friend I was not able to connect the scale. 

Currently I can't establish a bluetooth connection and can't find the raspberry via bluetooth. Open trickler app: You are not connected to a device.

Is it necessary to connect to a scale first?

**Comment by matsyngvelindgren** [comment]
IT WORKS SUPER NICE !!!
Awesome job man!!!
Many thanks and a big hug :-D

**Comment by erichiggins** [comment]
@GTMux From the Android App, you need to connect to the unit. You do **not** need to pair with the app from your Bluetooth menu/system settings.

- Open the menu (top-left corner of the screen)
- Select "Bluetooth Devices"
- Tap the button on the bottom-right of the screen and it should try to connect
- On successful connection, you should see the readings of the scale on screen
- Then, go back to the home view and you can control the trickler

I'm pretty sure I demonstrated all of this in [my YouTube video](https://www.youtube.com/watch?v=1BODqGHfVtk).

Hope this helps!

FWIW: The UI needs some updates to make this all easier based on what we've learned from using it.

**Comment by GTMux** [comment]
@erichiggins Thanks for the information but this is what I did. The app isn't showing any device. Of course I tapped the button on the bottem-right of the screen a couple of times. :) Maybe because of my old Android Tab. I will try with a newer model in the following days.

![IMG_8405](https://user-images.githubusercontent.com/56837063/71116503-92544100-21d4-11ea-8345-e2da90f35c97.JPG)

**Comment by erichiggins** [comment]
@GTMux That's interesting. A few things come to mind:

The Pi won't advertise itself on Bluetooth if it cannot detect the scale. The scale has to be plugged into it, turned on, and correctly configured.

Assuming a correct setup on the Pi/Scale, the question is: Is the problem the Pi or the Android app?
There are a few ways to find the answer to this. There's a free Android app called [nRF Connect](https://play.google.com/store/apps/details?id=no.nordicsemi.android.mcp&hl=en_US) which is pretty useful for debugging Bluetooth devices. It will let you see every device that is advertising.

**Comment by cyclepath37** [comment]
I was having the same issue as GTMux and then noticed only the boot partition was on the sd card. I’m not sure what i am doing wrong.

**Comment by erichiggins** [comment]
@cyclepath37 Which operating system are you using? Windows and Apple computers won't be able to see all three, but the `boot` and `code` should both be visible.

As long as your using the balena etcher program to flash the SD card, it should create the same result.

Did you configure your scale using the instructions listed in my article?

**Comment by cyclepath37** [comment]
Windows and using balena etcher as suggested. Only showing boot if i etch with or without decompressing. 
I have an A&D HR202i, but have it configured the same as you show with the fx120.

**Comment by erichiggins** [comment]
@cyclepath37 I'm not familiar with that particular model, so you may want to double check the manual to verify that it's RS232 interface operates the same way. For the sake of this conversation, let's assume it does for now.

Potential causes:
- Incorrect serial to USB adapters (they aren't all the same, even with identical physical connections)
- Scale not plugged in and turned on before booting the Pi
- Driver on the Pi being flakey. I've seen this on occasion and there's not a clear software workaround. Just have to reboot the Pi and try again

I can share another image tomorrow that includes logging, which would help to diagnose these problems more easily.

With the image you have, you can configure it to connect to your WiFi then SSH into the Pi and see what's going on, if you're comfortable working with the Linux command line.

**Comment by Shooter0423** [comment]
@erichiggins thanks for the updated image. I loaded to two different RPi Zero W's with different brand SD cards (1 16g, the other 32g) both connected to the scale  and worked to ready the scale value in the startup are. My issue comes in 2 parts.
1) After the initial success with the 2 setups, they developed a inability to connect, it was getting late so I shut everything done. tried it just now and things are working fine again, don't know what I did different? Is there a correct way to disconnect
2) Havent been able to get the vibro motor to do its thing. I know the motor and tube work cause I can test the on an Arduino. Is there a reason that the control pin changed, don't think it is the problem but just asking

**Comment by erichiggins** [comment]
@Shooter0423 Glad you were able to get them working fairly easily -- now to see if we can figure out some of the issues you've run into.

Let's start with the motors not running. The reason I changed the control pin from 15 to 12 was to allow the use of PWM (Pulse Width Modulation) control on the motors.

Here's a [pin out diagram for the Pi Zero W](https://pinout.xyz/pinout/io_pi_zero). Note that physical pin 15 is BCM 22 and physical pin 12 is BCM 18 (PWM 0)

You can check if this is the issue you're having pretty easily with two different methods:

First, power up the OpenTrickler Pi & scale in the usual way, connect w/ the app, then set a charge weight (like 10.00 gr) and hit the Play button. At this point, the motor should be running. If it's doing nothing, you can check the control pin on the Pi using a multimeter/voltmeter. You should see it turning on/off with approx 3V. If you don't, then check the other pin.

The other method you can use is to change the pin setting in the config file. Pull the microSD card out of the Pi and connect it to your computer. You should see a drive called `code`. On that drive, there will be a folder named `opentrickler` which has a file named `ecosystem.config.js`. Open that file with a text editor (NOT a word processor). You can change the `MOTOR_PIN` setting from `12` to whatever you are using. You may still want to use a multimeter in the same way I described above to verify that it's working as expected.

As for the issues you're having with disconnecting... I may need more info for that. There could be a software bug that I can fix. For now, how about we get the motor working first?

**Comment by Shooter0423** [comment]
Thanks for the quick response. I had debugged the issue the way you described but the real culprit turned out to be a bad solder joint. So everything is up and running. The Bluetooth connection seem to need more time than I’m allowing, I think that will eliminates the problem, I’ll let you know.

**Comment by erichiggins** [comment]
@Shooter0423 Great! If you are able to reproduce the bluetooth issues, go ahead and open a new issue here on GitHub so we can track it separately.

@GTMux @cyclepath37 Have either of you had luck getting yours working?

**Comment by GTMux** [comment]
@erichiggins: thanks for the update. As mentioned I‘m still waiting for the scale. When the scale arrives back at my home I‘ll let you know directly.

**Comment by cyclepath37** [comment]
Still stuck. I confirmed the fx & hr202i balance RS232 pins are the same. Does it need the entire breadboard motor circuit connected for the bluetooth to work? I am using an old dandy trickler and haven’t tied it in yet, so i have been trying to connect with just the bare pi zero without the motor circuit.

**Comment by erichiggins** [comment]
@cyclepath37 The Pi doesn't try to detect the motor circuitry at all (not even sure if that's possible), so you should be fine without it. The Pi only waits for a scale to be detected and the bluetooth drivers before it advertises itself so that the app can connect to it.

Did you [follow these instructions to configure your scale](https://blog.ammolytics.com/2019-04-30/diy-smart-trickler.html#prerequisite-scale-configuration)? Without those steps, the Pi won't be able to communicate with the scale. It's not just as easy as plugging it in, unfortunately.

**Comment by Datakalu** [comment]
Having the same result as GTMux and cyclepath37. Flashed the test image using Windows & balena, fx120 connected and configuration confirmed. Pi does not advertise on Bluetooth.

**Comment by Shooter0423** [comment]
Seems like I'm having a little better go at it then @GTMux,  @cyclepath37 and @Datakalu. I'm working with 2 different RPi Zero W's, one has been problematic from the beginning in regards to the Bluetooth connection. #1 would connect on startup only some of the times, like 1 in 10 on average. #2 will connect on startup most of the time, but sometimes it may take 2 or 3 startups (once as many a 5 this evening). I haven't isolated anything in my work area that could causing interference, (for a short time both where working kinda consistently and then my wife would come in the room and they would quite, I sure was going to miss her).
My startup procedure and some observations.
I quite plugging the power in and out on the pi and use the switch on a power strip, not sure how robust the micro usb sockets are.
I wait 2 minutes after turning on power to the pi to boot, I set a timer.
I then start the mobile app, wait a little, go into the menu and get Bluetooth Devices, wait a little, then hit the green button.  If it doesn't come back with the info from the scale in 10 or 15 seconds, it not going to because there is no connection ( I always wait for the app to tell me so ) try the green button a time or 2 more (never seems to work a second time if it didn't work the first) then turn off power, check the scale, check the cables, check that i'm not going crazy and do another startup.
I know its frustrating, but this thing works.
It looks like @erichiggins is using a stick powder in the video and I am working with load of Tac under a 77gr SMK, (also got a Varget load for a 6BR to work on) I haven't loaded real ammo yet, probably do that tomorrow as it looks like Wednesday will be the best day to go out this week.  I built my vibro-motor different so I'm tuning the motor.js and probably will need to reimplement some speed steps that have been commented out in trickler.js.  I'll start new threads with this info as I get it organized.

**Comment by chaseerickson12** [comment]
@erichiggins I’m having the same problem as everyone above. @Shooter0423 does the pi require a 16 or 32g sd? I’ve tried a 4 and a 8g and can’t get anything to work. I’m going to pick up a 16g after work and a usb to micro adventure adapter and see if that makes a difference over using a micro sd to sd adapter. Is the any way to test to make sure your on the right trail.

**Comment by Shooter0423** [comment]
@chaseerickson12 I was trying to run 2 pi's. One has a 16g standard Kingston, the other a 32g Ultra Scandisk. I swap the cards back and forth in the good working pi, as they both have the new image as supplied by @erichiggins recently. I use the adapter that comes with a micro card, have several laying around, don't think that matters. Anyway in my case it doesnt matter. Are you downloading the compressed image and using Belena to flash it to the card? Earlier when I was having problem with the first pi, a site suggested formatting cards with a freebie called 'SD Card Formatter', don't know that it actually fixed any problems but I use it every time before flashing cards now. Another suggested freebie, same site I think, is Note++. Works will when modifying/editing files in the code folder

**Comment by erichiggins** [comment]
Thanks to everyone for keeping me posted -- I know it can be frustrating but it really does help a lot to get feedback like this so that I can focus on the right problems that will work for everyone. Also thanks to @Shooter0423 for chiming in, helping others, and sharing your experience!

I think the biggest issue is that **we're flying blind**.

The only debugging info you have available is whether or not the app can connect to the Pi over bluetooth. There are a lot of things that could go wrong to prevent that, but without having some way to see that, it's impossible to find the root cause.

I suspect that if I can make the Pi's logs available via your web browser at http://opentrickler.local on your local Wifi, then we'd at least have a better idea of what's going right and what's going wrong, like:

- If the Pi is booting up at all.
- If the Bluetooth drivers are failing (I've seen this)
- If the Open Trickler program (on the Pi) is crashing and not restarting
- If the Pi isn't seeing the scale.

The other issue that I think may be worth addressing is allowing the Mobile app to connect to the Pi **even if the scale isn't connected**. This requires work on the mobile app as well, so it's a bit more work to pull off -- because of this, I've been kind of avoiding it.

I'd like to get everyone's thoughts on this though before I proceed. You might have better ideas and I'd love to hear them!

**Comment by erichiggins** [comment]
@chaseerickson12 A 4GB or 8GB card should be fine. It would probably even work on a 1GB card, though I haven't tried it.

**Comment by Shooter0423** [comment]
Thanks @erichiggins, glad to help. I think viewing a log will pay dividends faster. The mobile app is working good enough.

**Comment by Datakalu** [comment]
I have the wrong USB to Serial cable. I suspect that’s my problem and I’ll let you know how it plays out when I have the correct cable.

**Comment by chaseerickson12** [comment]
@Shooter0423 I have not been decompressing the file. Ill give it a shot. Im interested to hear if @Datakalu serial cable was the problem

**Comment by chaseerickson12** [comment]
https://www.amazon.com/gp/product/B075XHWVSJ/ref=as_li_qf_asin_il_tl?ie=UTF8&tag=ammolytics0f-20&creative=9325&linkCode=as2&creativeASIN=B075XHWVSJ&linkId=7c662ec9d4021bf3c1374f86ff1b9b0d 

Do we need to be using a usb to serial that requires this adapter? The one I purchased from the recommended list didn’t require this to be plugged in. Or could this be the problem that I’m experiencing

**Comment by erichiggins** [comment]
@chaseerickson12 Yes, you do need that actually. I know that you can connect the cables without it, but the adapter doesn't change the male/female interface, it changes the pin wiring arrangement. The A&D scales use a rather strange wiring setup which requires this in order to work.

**Comment by chaseerickson12** [comment]
@erichiggins i think I’m slowly eliminating my problems one at a time lol 😂. I’ll report back now that I’m trying the adapter

**Comment by chaseerickson12** [comment]
@erichiggins i can get Bluetooth connection 👍🏻 Now time to figure out where my wiring is messed up lol so the trickler works. If any of you used a mini solderless breadboard if you can post a photo that would be appreciated. I’m not 100% on if I followed the diagram on the main directions correctly.

**Comment by erichiggins** [comment]
@chaseerickson12 let's follow up on that over email tomorrow. Send me photos of what you have and I'll see if I can help sort it out.

**Comment by erichiggins** [comment]
# Update
I've made the logs available via web browser on the OpenTrickler with the following uSD image for the Pi. After you flash the uSD card with the following image, you will need to modify the `boot / wpa_supplicant.conf` file with the details for your WiFi network so that the Pi can connect to it.

Once booted, you can view the logs by visiting [`http://opentrickler.local/`](http://opentrickler.local/) in the web browser of any computer or mobile device that's on the same WiFi network. If it does not resolve, the most likely issue is a misconfigured `wpa_supplicant.conf` file.

**Do not share the contents of your `wpa_supplicant.conf` files here, as they will contain the password for your WiFi network and it may compromise your network security.**

Documentation for `wpa_supplicant.conf` can be found online. [Here's one example](https://linux.die.net/man/5/wpa_supplicant.conf).

## uSD Pi Image
This image is considered alpha quality and is intended only for testing purposes until these issues can be resolved and verified by all interested parties. Afterward, I'll provide some cleanups and a new image, which everyone should expect to download and reflash their uSD cards with.

[Download `opentrickler-1.1.6-ALPHA.img.zx`](https://drive.google.com/open?id=1H1-Y0sm0sSx40kNqHaOhjIgBHQxGmJH6)

## Checksums
Optional, in case you want to verify the image after downloading and extracting
- `sha  bf78327d5f9768637e6877cc8da8ccf3efe56fd7  opentrickler-1.1.6-ALPHA.img`
- `sha256  3bc9e5bb5a02d10dd160d68631198c8a7edd073a952ca8abe481bcecbaa8c797  opentrickler-1.1.6-ALPHA.img`
- `md5  149cb844703f2bec4e954ef5296bf449  opentrickler-1.1.6-ALPHA.img`

## Successful OpenTrickler Log Example

```
Setting: process.env[ NODE_ENV ] =  production
Setting: process.env[ MOTOR_PIN ] =  12
Setting: process.env[ SCALE_BAUD_RATE ] =  19200
Setting: process.env[ SCALE_DEVICE_PATH ] =  /dev/ttyUSB0
Setting: process.env[ DEVICE_NAME ] =  Trickler
===== STARTING UP =====
setting speed from undefined to {"true":60,"false":150}
Opening trickler...
Serial port open
weight: 0, unit: null, status: null, stable: false, stableTime: 0
weight: 0, unit: 0, status: null, stable: false, stableTime: 0
Scale is ready!
Trickler is ready! true
Scale weight reads: 0 0, stableTime: undefined
weight: 0, unit: 0, status: 0, stable: true, stableTime: 8
on -> stateChange: poweredOn
Scale is ready!
weight: 0, unit: 0, status: 0, stable: true, stableTime: 82
scale: undefined
Checking if ready... TRICKLER: true SCALE: true BT: true
on -> advertisingStart: success
advertising services...
```

## Unsuccessful OpenTrickler Log Example

```
Checking if ready... TRICKLER: false SCALE: false BT: true
weight: null, unit: null, status: null, stable: false, stableTime: 0
scale: undefined
Checking if ready... TRICKLER: false SCALE: false BT: true
weight: null, unit: null, status: null, stable: false, stableTime: 0
scale: undefined
Checking if ready... TRICKLER: false SCALE: false BT: true
```

**Comment by Datakalu** [comment]
Haven’t flashed the latest image but I’ve install the correct serial cable and was able to connect on a couple of occasions. However, it drops the connection before I can run the motor circuit.

**Comment by Datakalu** [comment]
Motor circuit runs, turns out it was me when returning to the main screen. Haven’t thrown powder yet but works as described. Bluetooth connection is hit and miss.

**Comment by erichiggins** [comment]
@Datakalu The logging capability in the latest image should give more insight as to why you may be having Bluetooth connection issues. I've seen a couple occasions where the USB/Serial driver acts strangely.

**Comment by Datakalu** [comment]
@erichiggins I’ve never played around with tech projects like this and it’s been an education getting to this point. I’ll keep slowly chipping away at it but I’m learning as I go. Great project and support, thank you!

**Comment by MrMr-No** [comment]
Can we get  Pi password?

**Comment by erichiggins** [comment]
@MrMr-No Yes! I actually shared it in [an earlier post on this thread](https://github.com/ammolytics/projects/issues/28#issuecomment-566235134).

**Comment by cyclepath37** [comment]
Update: I continued to struggle with bluetooth connection loss as many others had also. It would connect, then drop and sometimes take several attempts to reconnect, followed by dropping it again. I ended up ordering a different serial cable. 
It must have been the cable because the new cable has connected right away with no loss after running it for 20 minutes!!!!
Here is the cable I ordered 
USB to RS232 Adapter with PL2303 Chipset, CableCreation 3.3 ft USB 2.0 to RS232 Female DB9 Serial Converter Cable for Cashier Register, Modem, Scanner, Digital Cameras, CNC, 1M Black https://www.amazon.com/dp/B0769FY7R7/ref=cm_sw_r_cp_api_i_EHphEbVHZDVP9
Up and running!!!!

**Comment by Shooter0423** [comment]
Update:
I have been having intermittent connection problems. Most of the time the Bluetooth would connect but the trickler program and the scale would not report TRUE and the program would stay in a loop and never connect. I would cycle the power to the RPi and let the program start over.  Generally after 2 or 3 tries it would connect and all would be good. I followed @cyclepath37 cable suggestion from above, (I did not replace the USB to micro USB adapter) and been starting and restarting over the last couple hours, probably 10 or 12 times by now. The RPi has connected everytime.

**Comment by erichiggins** [comment]
@Shooter0423 Interesting! Thanks for the update! I'll order one of these cables and try it out, then update the parts list. So happy that the community was able to find a great solution for this one!

**Comment by Shooter0423** [comment]
@erichiggins 
I think the cable definitely fixed my connection issues. More starts and restarts today, working every time. I have the led code to monitor startup embedded in the Index.js file and the led's will even turn on in different orders as the program is retrying to establish connections. Interesting situation, if you forget to turn on the scale, the BT will connect, program and scale will report FALSE, turn on the scale and both will report TRUE and OpenTrickler is ready to roll.
I need to post something about the edits to the code and layout on the proto-bonnet. A picture will not be very good, and a hand sketch probably worse, have any suggestions. And what to do with the modified files.
![ProtoBonnet1](https://user-images.githubusercontent.com/57802826/73711919-0f2c7300-46ce-11ea-8314-db5b98cdeb15.jpg)
Pic show a drv8825 stepper driver, i got that mostly working, not sure best place to activate the powder measurer, looking at the code in trickler.js where it starts the motor when the scale is zero for more than 1 second

**Comment by Datakalu** [comment]
I ordered the Cable Creations cable as well and I am no longer experiencing Bluetooth issues. So far it’s been 100% and connecting quickly. Thank you @cyclepath37 & @Shooter0423 !!! Moving forward and starting to build a Dandy style trickler for the system.

**Comment by Shooter0423** [comment]
@Datakalu great news, haven't heard any news from @erichiggins about his experience with the new cable. We owe a bit of thanks to @cyclepath37. If you start a new thread about your trickler build, i'll post a couple pictures of a trickler I built out of gray pvc conduit connectors (iirc about $5 maybe $7) and eric's vibro-motor circuit

**Comment by MrMr-No** [comment]
I  activate stepmotor only when I need need more then 30 grains 

 case (tickDelta > 1499):
        this.motor.speed = SPEEDS.VERY_FAST
        
...
....

drv8825 working very well :)

**Comment by erichiggins** [comment]
I ordered a new cable, but I'm out of town for work and won't get a chance to try it until tomorrow.

Thanks to everyone for sharing their feedback & suggestions, generally helping each other out and being awesome!

**Comment by erichiggins** [comment]
@Shooter0423 Two ways to share your code changes:

1. You can create a public Gist here on GitHub. Very easy.
  https://gist.github.com/
2. You can create a [Pull Request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests) to add your changes into the repo. I can review it, suggest changes, then merge it.

Don't hesitate to email me if you need some help.

**Comment by Shooter0423** [comment]
@MrMr-No, great idea to put it in as another speed option.  I have struggled to keep it from double throwing in shouldGo or startWhenReady.  Are you setting up a function for the stepper code and calling it and are you using setIntervals or msleep/usleep.  My coding skill with js is poor, any code you want to share will be greatly appreciated.

**Comment by MrMr-No** [comment]
From case i calling pythonshell cod and using python to run step motor. First check if requested weight is more then 30 grains then and scale is set to 0 . 

case (tickDelta > 1499 && this.scale.weight === 0  ) :
        
     
      let {PythonShell} = require('python-shell')
PythonShell.run('/home/pi/projects/trickler/peripheral/lib/steper.py', null, function(err) {
  if (err) throw err
  console.log('finished stepper')

**Comment by erichiggins** [comment]
@cyclepath37 I got the new cable  -- are you using it with the adapter, or just connected directly to the scale?

**Comment by cyclepath37** [comment]
@erichiggins I use the null modem adapter, and I ended up getting a 90degree adapter also since i have my scale pushed up against the wall. StarTech.com Right Angle DB9 to DB9 Serial Cable Adapter Type 2 - M/F (GC99MFRA2) https://www.amazon.com/dp/B001S2PYV4/ref=cm_sw_r_cp_api_i_rVhqEbJXQ41NM
![82D06728-983F-4016-BE3D-9331830155A9](https://user-images.githubusercontent.com/58094393/74110504-1e427380-4b5b-11ea-8043-7dd543a929cc.jpeg)
I also wired in the push button from the dandy. This is allowing me to manually add one kernel if needed. I typically set the app to .02gr below my target. Had to adjust the values a bit since I am using a modified forster drop tube(long).

**Comment by erichiggins** [comment]
@cyclepath37 Great setup -- thanks for the photo!

**Comment by erichiggins** [comment]
I've had a chance to try the new cable and it worked great for me! I'll try doing a few rapid reboots today just to verify that it's more stable, but so far so good.

**Comment by Passaita** [comment]
Hello everyone, first of all thank you for moving this ahead. Really great addition to our hobby.
I am waiting for my scale to be delivered and decided to pre-build trickler in advance. A couple attempts to build the code on raspbian using installation script failed (mostly due to my bad internet some packages wouldn't download) so I installed the image.
After reading the thread looks like Bloetooth is getting more stable with certain cable. At the same time I am wondering:
- Can weight selection be implemented on the web interface?
- Since PRi has SD card some presets (weight) can be saved too.

I can contribute my 3D design skills to this project. Once my system is running I am planing to make a case for RPi and all the components. Anything else you were thinking about?

**Comment by Shooter0423** [comment]
@Passaita welcome to the group. You are correct about the cable, definitely buy the cable @cyclepath37 suggested, im not sure if @erichiggins update the list of materials.
The Bluetooth is very stable, the old cable was causing a connection error between the RPi and the scale, but the new cable seems to have resolved that problem.
I'm in the process of testing an "auto-throw" feature using a Lee Deluxe Power Measure activated by a stepper motor contolled by the RPi. If you haven't built the expansion board for the RPi, i suggest that you use the proto bonnet that eric suggest, it make a very clean setup. The stepper drive I use will fit on the unused area if you keep the vibro control footprint kinda small. There is a pic in a post i made earlier to give an idea of the layout needed for the driver.

**Comment by Passaita** [comment]
I don't have a powder measure to play with and to be honest I am currently 3D designing powder trickler just for this project as I didn't have one. I know, I know, I am sorry, no comments :)
Stepper motors are great and I used them in the 3D printer but the complexity with drivers scares me. Maybe because I have a lot of experience in RC planes and prefer servos I built controllers for my blinds at home using ESP8266 ESP-12 NodeMCU WeMos and servos without any additional drivers. This approach could be much simpler for others to follow and cheaper rather than stepper motors.

**Comment by Passaita** [comment]
My scale arrived today! Will try to connect everything tonight.

**Comment by Passaita** [comment]
> My scale arrived today! Will try to connect everything tonight.

It works!!!
As described above pin was changed from 15 to 12 and pre-built image works.
Dispensing is way too slow but it is most likely because of my homemade trickler. Maybe pipe is too long, I made it so it can reach the center of the scale. Or vibro motor that I could source at a local shop isn't strong enough.
Anyway even with really slow dispensing on the last stage of loading code doesn't pause long enough to wait for kernels to saddle down and this is typically causing an overload 0.02 - 0.04 grains.
Another thing I have noticed (again this is mostly because of the pipe) on the fast stage of dispensing vibrations are causing kernels to almost plug the pipe. I manually added a couple grains and motor changed vibration frequency. This change made kernels move in the previously "plugged" pipe. So my suggestion would be to use various vibro modes for the fast stage.

**Comment by Shooter0423** [comment]
@Passaita good to hear that thing worked well. I had some of the same issues with the trickler. I built a trickler out of parts on hand but did use the vibro motor listed in the BOM. I rewrote the portion of code that @erichiggins had developed. Eric had set me up a discussion thread for PWM control, I didn't have much success in that area but that thread might be a place to continue this discussion and keep this thread focused on installation. By the why, what powder are you using, i'm using Tac and Varget, a ball powder and a stick powder.  You said the pipe may be to long, the pipe I built is 6" long made of the aluminum tube spec'd in the build list. My thought was the hole I drilled may be to big and causing the pipe to clog. And lastly, when you say it to slow, what amount of time are you referring, with Tac I get a charge at the target weight in 20 secs 95% of the time. I load in blocks of 100 as a match is 69 rnds or 88 rnds (20 rnds for record x3 or 4 and sighters as allowed) so if the throw is +/-0.02gr it gets used but put in the front row and get used for foulers or something not in the match.

**Comment by Datakalu** [comment]
![2113FD5D-4D19-43A0-BE23-D77FD1D50156](https://user-images.githubusercontent.com/59381352/79692789-52ee3d80-8224-11ea-8021-1ec00021f65a.jpeg)
I put together a knock off of a Dandy Trickler that when tuned to the powder with the dampening weight it will hit the target weight spot on every time with either ball or stick propellants.
I’ll throw a bulk charge to within a 1/2 grain or so and I charge cases just about as quickly as I move the pan around.
The system rocks, thank you all!!!!!!!

**Comment by Passaita** [comment]
> @Passaita 

- PWM control, I didn't have much success in that area but that thread might be a place to continue this discussion and keep this thread focused on installation. _Absolutely agree and will continue there. This thread seemed to be the most live one? Looking forward to have a look at it. I have no experience with JS but some basic knowledge in Python._
- What powder are you using, i'm using Tac and Varget, a ball powder and a stick powder. _I am using Alliant Reloder 17. Never worked with other powders this one looks lite cuts of little rods._
- You said the pipe may be to long, the pipe I built is 6" long made of the aluminum tube spec'd in the build list. _Mine is stainless steel 3/8 OD I had in the garage._
- My thought was the hole I drilled may be to big and causing the pipe to clog. _I experimented a little: mirror polished inside and drilled a little bigger hole, now the flow is a visually better_
- When you say it to slow, what amount of time are you referring? _It takes over two minutes for 40.3 grains. I was faster using hands :) But it's project in process_

There is following response from **Datakalu** and I want to say that I have made similar modifications. I have moved the motor instead of the weight and it gives different results on flow, speed and accuracy.

![vibromove](https://user-images.githubusercontent.com/25120690/79695875-7326f800-8236-11ea-8280-75e279184bc2.PNG)

**Comment by Datakalu** [comment]
Brilliant!

**Comment by Shooter0423** [comment]
@Datakalu very nice looking trickler
@Passaita same with yours
Reloder 17 probably meters a lot like varget, 
I forgot to add in my earlier post that I throw the bulk of the charge to within 0.5gr with a powder measurer then let the trickler finish. If I where to try to trickler the entire charge (approx. 24 gr) it would take a long time. I'm thinking to get an efficient speed you need to throw the bulk of the charge and then trickle to the target.

**Comment by erichiggins** [comment]
Hey everyone! I've updated the instructions on the main [README](https://github.com/ammolytics/projects/blob/develop/trickler/peripheral/README.md) to reflect the work that was done here to use ready-to-flash images. I've also added links to two new images: production and dev.

### Notable changes:
- The `ecosystem.config.js` file now assumes you've soldered motor control to PIN 12. If you're still using PIN 15, you can update the file on your `CODE` partition after flashing the latest image.
- The production image removes SSH access to reduce the the security risks I mentioned above.
- The log files viewed from http://opentrickler.local now show the entire log file, not just starting with the last 10 lines.
- Windows-friendly partitions (`BOOT` and `CODE`) have been changed to uppercase to resolve issues with older versions of Windows which may not be able to view them.
- Some issues with Bluetooth advertising have been fixed. If you've had trouble with the app not being able to find your Open Trickler, this new image may help.

With that said, I think this ticket can finally be closed out. Please feel free to open a new ticket if you experience issues with the latest images!

**Comment by erichiggins** [comment]
Hey everyone! I created a Discord to make it easier for folks who are building an Open Trickler to chat with and support each other.

https://discord.gg/Bx3ynT

This invite link expires in 1 day to keep out spammers, but feel free to email me directly and I'd be happy to invite you!
